rustler/src/types/primitive.rs

```diff
@@ -60,3 +59,34 @@ impl<'a> Decoder<'a> for bool {
         atom::decode_bool(term)
     }
 }
+
+impl Encoder for f32 {
+    fn encode<'a>(&self, env: Env<'a>) -> Term<'a> {
+        #[allow(clippy::cast_lossless)]
+        unsafe {
+            Term::new(
+                env,
+                rustler_sys::enif_make_double(env.as_c_arg(), *self as f64),
+            )
+        }
+    }
+}
+impl<'a> Decoder<'a> for f32 {
+    fn decode(term: Term) -> NifResult<f32> {
+        #![allow(unused_unsafe)]
+        let mut res: f64 = Default::default();
+        if unsafe {
+            rustler_sys::enif_get_double(term.get_env().as_c_arg(), term.as_c_arg(), &mut res)
+        } == 0
+        {
+            return Err(Error::BadArg);
+        }
+        let res = res as f32;
```

Comment:
I don't know enough about the behaviour of `as` w.r.t. floating point numbers. I guess this results in a potential precision loss and not only a loss in the range of maximum values? Is this something users would be aware of?

Reply:
It seems that converting from f64 to f32 leads to precision loss, but:

1. The same attitude is ALREADY presented in `rustler` (but generated by macro).
2. Probably, library user knows for himself better WHY he's trying to use `f32` instead of `f64` while erlang provides all the floating point values as `doubles`. The only reason can be "to speed up a math", and for that case it seems to be a reasonable tradeoff.

As an alternative, possible solution will be totally remove `f32` as a type that implements `rustler::{Encoder, Decoder}` traits.

Reply:
Yeah, I think you are right, users should be aware of this already.